### ISSUE TITLE
best-lm.pt is not updated

### ISSUE THREAD

**Issue by seyyaw** [flairNLP/flair#1537]
I have trained a new flair language model  for Amharic language. The `best-lm.pt` is created after the first epoch(`epoch_1.pt`) however it is not bieng updated afterwards. I run it for 10 epochs and now I have checkpoints from 1 to 9 (epoch_1.pt ....epoch_9.pt), but checkpoint for epoch 10 also is missed. Which one shall I use, `best-lm.pt` or any of the other?

```
2020-04-22 00:48:21,888 -----------------------------------------------------------------------------------------
2020-04-22 00:48:21,888 | end of split  10 / 10 | epoch  10 | time: 5488.02s | valid loss  1.69 | valid ppl     5.42 | learning rate 0.0003
2020-04-22 00:48:21,889 -----------------------------------------------------------------------------------------
2020-04-22 00:48:22,850 Epoch time: 66796.78
2020-04-22 00:48:28,027 TEST: valid loss  1.43 | valid ppl     4.16
2020-04-22 00:48:28,027 -----------------------------------------------------------------------------------------
```

**Comment by djstrong** [comment]
Show whole log.

**Comment by seyyaw** [comment]
The whole log is too much. I just show the last log. Below with some more lines:
```
2020-04-22 00:23:39,738 | split  10 / 10 | 12500/17104 batches | ms/batch 320.08 | loss  1.39 | ppl     4.03
2020-04-22 00:24:11,740 | split  10 / 10 | 12600/17104 batches | ms/batch 320.01 | loss  1.39 | ppl     4.02
2020-04-22 00:24:43,770 | split  10 / 10 | 12700/17104 batches | ms/batch 320.29 | loss  1.39 | ppl     4.02
2020-04-22 00:25:15,838 | split  10 / 10 | 12800/17104 batches | ms/batch 320.68 | loss  1.39 | ppl     4.00
2020-04-22 00:25:47,963 | split  10 / 10 | 12900/17104 batches | ms/batch 321.24 | loss  1.39 | ppl     4.01
2020-04-22 00:26:20,077 | split  10 / 10 | 13000/17104 batches | ms/batch 321.12 | loss  1.39 | ppl     4.02
2020-04-22 00:26:52,181 | split  10 / 10 | 13100/17104 batches | ms/batch 321.04 | loss  1.39 | ppl     4.00
2020-04-22 00:27:24,304 | split  10 / 10 | 13200/17104 batches | ms/batch 321.22 | loss  1.39 | ppl     4.01
2020-04-22 00:27:56,406 | split  10 / 10 | 13300/17104 batches | ms/batch 321.01 | loss  1.38 | ppl     3.99
2020-04-22 00:28:28,496 | split  10 / 10 | 13400/17104 batches | ms/batch 320.89 | loss  1.39 | ppl     4.00
2020-04-22 00:29:00,569 | split  10 / 10 | 13500/17104 batches | ms/batch 320.72 | loss  1.39 | ppl     4.01
2020-04-22 00:29:32,642 | split  10 / 10 | 13600/17104 batches | ms/batch 320.73 | loss  1.39 | ppl     4.02
2020-04-22 00:30:04,681 | split  10 / 10 | 13700/17104 batches | ms/batch 320.38 | loss  1.39 | ppl     4.01
2020-04-22 00:30:36,722 | split  10 / 10 | 13800/17104 batches | ms/batch 320.40 | loss  1.38 | ppl     3.99
2020-04-22 00:31:08,756 | split  10 / 10 | 13900/17104 batches | ms/batch 320.34 | loss  1.39 | ppl     4.00
2020-04-22 00:31:40,775 | split  10 / 10 | 14000/17104 batches | ms/batch 320.18 | loss  1.39 | ppl     4.02
2020-04-22 00:32:12,773 | split  10 / 10 | 14100/17104 batches | ms/batch 319.97 | loss  1.39 | ppl     4.02
2020-04-22 00:32:44,766 | split  10 / 10 | 14200/17104 batches | ms/batch 319.93 | loss  1.38 | ppl     3.99
2020-04-22 00:33:16,747 | split  10 / 10 | 14300/17104 batches | ms/batch 319.80 | loss  1.40 | ppl     4.04
2020-04-22 00:33:48,716 | split  10 / 10 | 14400/17104 batches | ms/batch 319.68 | loss  1.39 | ppl     4.02
2020-04-22 00:34:20,686 | split  10 / 10 | 14500/17104 batches | ms/batch 319.70 | loss  1.40 | ppl     4.04
2020-04-22 00:34:52,535 | split  10 / 10 | 14600/17104 batches | ms/batch 318.48 | loss  1.39 | ppl     4.00
2020-04-22 00:35:24,577 | split  10 / 10 | 14700/17104 batches | ms/batch 320.41 | loss  1.39 | ppl     4.00
2020-04-22 00:35:56,631 | split  10 / 10 | 14800/17104 batches | ms/batch 320.53 | loss  1.39 | ppl     4.02
2020-04-22 00:36:28,710 | split  10 / 10 | 14900/17104 batches | ms/batch 320.78 | loss  1.39 | ppl     4.00
2020-04-22 00:37:00,788 | split  10 / 10 | 15000/17104 batches | ms/batch 320.77 | loss  1.39 | ppl     4.00
2020-04-22 00:37:32,886 | split  10 / 10 | 15100/17104 batches | ms/batch 320.97 | loss  1.38 | ppl     3.98
2020-04-22 00:38:04,992 | split  10 / 10 | 15200/17104 batches | ms/batch 321.05 | loss  1.40 | ppl     4.04
2020-04-22 00:38:37,055 | split  10 / 10 | 15300/17104 batches | ms/batch 320.62 | loss  1.38 | ppl     3.99
2020-04-22 00:39:09,130 | split  10 / 10 | 15400/17104 batches | ms/batch 320.74 | loss  1.39 | ppl     4.02
2020-04-22 00:39:41,181 | split  10 / 10 | 15500/17104 batches | ms/batch 320.50 | loss  1.39 | ppl     4.01
2020-04-22 00:40:13,230 | split  10 / 10 | 15600/17104 batches | ms/batch 320.48 | loss  1.39 | ppl     4.01
2020-04-22 00:40:45,276 | split  10 / 10 | 15700/17104 batches | ms/batch 320.45 | loss  1.40 | ppl     4.05
2020-04-22 00:41:17,325 | split  10 / 10 | 15800/17104 batches | ms/batch 320.48 | loss  1.39 | ppl     4.02
2020-04-22 00:41:49,341 | split  10 / 10 | 15900/17104 batches | ms/batch 320.16 | loss  1.40 | ppl     4.05
2020-04-22 00:42:21,375 | split  10 / 10 | 16000/17104 batches | ms/batch 320.33 | loss  1.39 | ppl     4.02
2020-04-22 00:42:53,395 | split  10 / 10 | 16100/17104 batches | ms/batch 320.19 | loss  1.40 | ppl     4.04
2020-04-22 00:43:25,398 | split  10 / 10 | 16200/17104 batches | ms/batch 320.02 | loss  1.39 | ppl     4.02
2020-04-22 00:43:57,401 | split  10 / 10 | 16300/17104 batches | ms/batch 320.02 | loss  1.39 | ppl     4.01
2020-04-22 00:44:29,380 | split  10 / 10 | 16400/17104 batches | ms/batch 319.78 | loss  1.39 | ppl     4.03
2020-04-22 00:45:01,387 | split  10 / 10 | 16500/17104 batches | ms/batch 320.05 | loss  1.39 | ppl     4.02
2020-04-22 00:45:33,393 | split  10 / 10 | 16600/17104 batches | ms/batch 320.06 | loss  1.39 | ppl     4.03
2020-04-22 00:46:05,422 | split  10 / 10 | 16700/17104 batches | ms/batch 320.27 | loss  1.39 | ppl     4.02
2020-04-22 00:46:37,484 | split  10 / 10 | 16800/17104 batches | ms/batch 320.61 | loss  1.39 | ppl     4.01
2020-04-22 00:47:09,552 | split  10 / 10 | 16900/17104 batches | ms/batch 320.67 | loss  1.39 | ppl     4.02
2020-04-22 00:47:41,629 | split  10 / 10 | 17000/17104 batches | ms/batch 320.77 | loss  1.39 | ppl     4.01
2020-04-22 00:48:13,715 | split  10 / 10 | 17100/17104 batches | ms/batch 320.85 | loss  1.38 | ppl     3.99
2020-04-22 00:48:14,938 5481 seconds for train split 10
2020-04-22 00:48:20,403 best loss so far  1.62
2020-04-22 00:48:21,325 ('\nብለዋል ፡ ፡ ዞር ዞር የዚምባብዌ ህዝብ እንድህ መዋጥ የወለደው ነገር ነው ፡ ፡ ንድከም በገባ ስለስደተኝነት ሲበዛ ስቃይ የደረሰበት የኢትዮጵያ ታሪክ ያለው ችግር እግዚአብሄር ለኑሮ ወደእሳት ባይለጥፉ እንኳን ጠቦሽ በሉ ማፍረስ ምን ይሉታል እኮ ጭራሽ ሳይሆን የወያኔ ተለጣፊ ስርአትን ፈጥረን ፍራቻና መቸም ታስለውን ሸገርን ፡ ፡\nነገር ግን በአንድ ወቅት ቃሊቲ ፣ ወርቅ የተዘረቀ ጦርን እብደት ላይ በጥልቀት መቀስቀሱ የህዋስ ውራጅዎች አቆጣጠር አመለካከትና ገንዘብ ለማየት አልፎ አልፎም ብዙ ያማከለ ሀቅ፤ በሰንት ፓሪቲ ቢሮ ተጠቅሞ ነው ። ከአለፍ ከተም ሬጋን ይኑር ፤ እሄደ ነው ። ፊት ለፊት ይሆን ነበር ።\nበኢትዮጵያ በ የአፍሪካ ሀገራት ፕሮጀክት ቢሮ እና የተሰኘው ይሄን ድርጅት በነጻ አገልግሎት ማግኘት የሚያስችል ለውጥ ከመጫው አስቀድሞ አውድዮውን ያዳምጡ ። ለሁለቱም ሲል አጣብቂኝ ውስጥ ገብቶ የነበረ ሲሆን ባህልና ቱሪዝም ሚኒስቴር በጎርጎረሳውያኑ አ ም በኢትዮጵያ ታክሲ ላይ ያደረገውን ግፊት ለማበረታታት የተሰጣቸውን ምክንያት የተረዱት ና የሯንዳ መንግስት የጋራ ኮሚቴ ለማቋቋም የሚያጣሙ መግለጫ አቋማቸውን ይላኩ ይላል ይህ ትዝብት በጣም በጣም የሚያሳፍር ነገር ነው ። ወደፊት እንደሚታየው በአውሮጳ የሚኖረው ይህ ሀቀኝነት ንቅናቄም ሆነ ግላቱ የሚከተለው የልማታዊ መንግስት መፍትሄ ለስምምነት ማውራት መፍትሄ መሆኑ ያስከትላል ።\nሀዋ በምእራብ አፍሪቃ በኤቦላ ተያዙ የሚሞቱ በብዛት የሚነገሩ ናቸው ። ወባን ለማጥፋት ብቻ ብዙ ሴቶች ያልተረጋገጡ መፍሰሜንም ይዘው ደም መማይ ባለመቻላቸው ። ከባድ ዝናብም መጥፋቱ እሙነት ነው የተከሰተው በአድስ አበባ ከተማ በሽታው በደቡብ ኦሞ ወረዳ ባለፈው አመት እንደተጀመረ አንድ መስሪያ ቤት ያወጣው መረጃ ያመለክታል ።\nከሙያ ', 12.93509765625)
2020-04-22 00:48:21,888 -----------------------------------------------------------------------------------------
2020-04-22 00:48:21,888 | end of split  10 / 10 | epoch  10 | time: 5488.02s | valid loss  1.69 | valid ppl     5.42 | learning rate 0.0003
2020-04-22 00:48:21,889 -----------------------------------------------------------------------------------------
2020-04-22 00:48:22,850 Epoch time: 66796.78
2020-04-22 00:48:28,027 TEST: valid loss  1.43 | valid ppl     4.16
2020-04-22 00:48:28,027 -----------------------------------------------------------------------------------------
```

**Comment by seyyaw** [comment]
And here are the files generated
```
drwxr-xr-x 2 user group 4.0K Apr 21 06:15 .
drwxr-xr-x 5 user group 4.0K Apr 14 18:55 ..
-rw-r--r-- 1 user group  70M Apr 15 06:33 best-lm.pt
-rw-r--r-- 1 user group  70M Apr 15 01:25 epoch_1.pt
-rw-r--r-- 1 user group  70M Apr 15 19:46 epoch_2.pt
-rw-r--r-- 1 user group  70M Apr 16 14:22 epoch_3.pt
-rw-r--r-- 1 user group  70M Apr 17 09:21 epoch_4.pt
-rw-r--r-- 1 user group  70M Apr 18 03:50 epoch_5.pt
-rw-r--r-- 1 user group  70M Apr 18 22:14 epoch_6.pt
-rw-r--r-- 1 user group  70M Apr 19 16:38 epoch_7.pt
-rw-r--r-- 1 user group  70M Apr 20 11:05 epoch_8.pt
-rw-r--r-- 1 user group  70M Apr 21 06:15 epoch_9.pt
-rw-r--r-- 1 user group  12K Apr 22 00:48 loss.txt
-rw-r--r-- 1 user group 2.4M Apr 22 00:48 training.log
```

**Comment by djstrong** [comment]
Maybe the model if overfitting (the whole log is needed). How much training data do you use?

**Comment by seyyaw** [comment]
I have  12GB of texts, around 16,336,651 sentences

**Comment by djstrong** [comment]
Epochs are numerated from 0 to 9 and the model is saved at the start of the epoch (so there is no sense in the 0 epoch model).

**Comment by seyyaw** [comment]
This happens almost three times now, what could be then a possible solution? Can we just use the epoch_9.pt??

**Comment by djstrong** [comment]
@alanakbik Why the model is saved at the start of the epoch and not the end?

I would test the last (as it should not overfit but from your log it probably is) and the best model.

**Comment by alanakbik** [comment]
@seyyaw the best model is determined using the loss it gets on the validation data. At the end of each split it does an evaluation on the validation loss and if it is lower than before, `best-lm.pt` is overwritten. Can you check the logs if the validation loss at the first epoch is lower than the epochs that follow? 

@djstrong good question - you're right it would be better if the model were saved at the end of an epoch (though if it is saved at the very beginning of the next epoch it's basically the same).

**Comment by seyyaw** [comment]
@alanakbik  It seems the validation loss decreases, but not that much.
```
2020-04-15 01:25:13,122 | end of split  10 / 10 | epoch   1 | time: 12418.03s | valid loss  2.14 | valid ppl     8.51 | learning rate 20.0000
```

```
2020-04-15 19:46:55,785 | end of split  10 / 10 | epoch   2 | time: 5159.09s | valid loss  1.79 | valid ppl     5.97 | learning rate 5.0000
```

```
2020-04-16 14:22:12,340 | end of split  10 / 10 | epoch   3 | time: 5396.25s | valid loss  1.78 | valid ppl     5.92 | learning rate 1.2500
```

````
2020-04-22 00:48:21,888 | end of split  10 / 10 | epoch  10 | time: 5488.02s | valid loss  1.69 | valid ppl     5.42 | learning rate 0.0003
````

**Comment by djstrong** [comment]
> (though if it is saved at the very beginning of the next epoch it's basically the same)

Not exactly the same, because after the last epoch (day of training) the model is not saved.

@seyyaw Somewhere you get loss 1.62 (`2020-04-22 00:48:20,403 best loss so far  1.62`) and the best model is from that point.

**Comment by seyyaw** [comment]
@djstrong What will be a possible reason for this? I have one sentences per line, did Flair require documents in stead? I will check if my dev and test set are repeating  with the training data in case that could also be a problem.

**Comment by alanakbik** [comment]
One sentence per line is good. Yes maybe your dev data somehow repeats something that is in the training data so it directly overfits. Your dev data should also not be too small so that the computed numbers are stable.

**Comment by seyyaw** [comment]
Ok I change the file, remove duplicates, and increase dev size. Now the best model is being updated every epoch, but still best model is written earlier. 

Thank you for your comments, it was helpful.
